R/utils-code_dependency.R

```diff
@@ -0,0 +1,350 @@
+      sym_cond <- sym_cond[!x[sym_cond, "text"] %in% x[sym_form_cond, "text"]]
+
+      object_ids <- x[sym_cond, "id"]
+      dollar_ids <- x[x$"token" == "'$'", "id"]
+      after_dollar <- object_ids[(object_ids - 2) %in% dollar_ids]
+      sym_cond <- setdiff(sym_cond, which(x$id %in% after_dollar))
+
+      if (length(sym_cond) >= 2) {
+        ass_cond <- grep("ASSIGN", x$token)
+        text <- unique(x[sort(c(sym_cond, ass_cond)), "text"])
+
+        if (text[1] == "->") {
+          rev(text[-1])
+        } else {
+          text[-1]
+        }
+      }
+    }
+  )
+
+  side_effects <- grep("@effect", pd[pd$token == "COMMENT", "text"], value = TRUE)
+  check_effects <-
+    if (length(side_effects) > 0) {
+      affected <-
+        unlist(strsplit(sub("\\s*#\\s*@effect\\s+", "", side_effects), "\\s+"))
+
+      unique(c(object_names, affected))
+    } else {
+      object_names
+    }
+
+  effects <- lapply(
+    check_effects,
+    function(x) {
+      maxid <- suppressWarnings(max(occurrence[[x]]))
+      return_code_for_effects(
+        x,
+        pd = calls_pd,
+        occur = suppressWarnings(lapply(occurrence, function(x) setdiff(x, maxid:max(maxid, max(x))))),
+        cooccur = cooccurrence,
+        eff = NULL
+      )
+    }
+  )
+  names(effects) <- check_effects
+
+  list(
+    occurrence = occurrence,
+    cooccurrence = cooccurrence,
+    effects = effects
+  )
+}
+
+#' @title Get children calls within `getParseData()` object
+#' @param pd `list` of `data.frame`s of results of `utils::getParseData()` trimmed to unique `parsed_code` calls
+#' @param parent parent id in `utils::getParseData()`
+#' @return Row `binded` `utils::getParseData()` of all calls.
+#' @keywords internal
+get_children <- function(pd, parent) {
+  idx_children <- abs(pd$parent) == parent
+  children <- pd[idx_children, c("token", "text", "id")]
+  if (nrow(children) == 0) {
+    return(NULL)
+  }
+
+  if (parent > 0) {
+    do.call(rbind, c(list(children), lapply(children$id, get_children, pd = pd)))
+  }
+}
+
+#' @title Detects `"SYMBOL"` tokens for row `binded` `getParseData()` structure
+#' @param object `character` containing the name of the object
+#' @param pd `list` of `data.frame`s of results of `utils::getParseData()` trimmed to unique `parsed_code` calls
+#' @return A `logical` vector pointing in which elements of `pd` the `SYMBOL` token row has `object` in text column
+#' @keywords internal
+detect_symbol <- function(object, pd) {
+  unlist(
+    vapply(
+      pd,
+      function(call) {
+        is_symbol <-
+          any(call[call$token %in% c("SYMBOL", "SYMBOL_FUNCTION_CALL"), "text"] == object) &&
+            !any(call[call$token == "SYMBOL_FORMALS", "text"] == object)
+
+        object_ids <- call[call$text == object, "id"]
+        dollar_ids <- call[call$"token" == "'$'", "id"]
+        after_dollar <- object_ids[(object_ids - 2) %in% dollar_ids]
+        object_ids <- setdiff(object_ids, after_dollar)
```

Comment:
this code can be separated and used in `cooccurrence <- lapply` also.


```suggestion
        is_symbol <-
          any(call[call$token %in% c("SYMBOL", "SYMBOL_FUNCTION_CALL"), "text"] == object) &&
            !any(call[call$token == "SYMBOL_FORMALS", "text"] == object)

        object_ids <- call[call$text == object, "id"]
        dollar_ids <- call[call$"token" == "'$'", "id"]
        after_dollar <- object_ids[(object_ids - 2) %in% dollar_ids]
        object_ids <- setdiff(object_ids, after_dollar)
```

Reply:
will continue on this one under this card https://github.com/insightsengineering/teal.code/issues/154